File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

public class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;

        Map<String, Integer> map = new HashMap<>();
        
        int len = friends.length;

        for (int i = 0; i < len; i++) {
            map.put(friends[i], i);
        }

        int[][] table = new int[len][len];

        int[] giftScore = new int[len];

        for (String gift : gifts) {
            String[] name = gift.split(" ");
            String giver = name[0];
            String getter = name[1];

            giftScore[map.get(giver)]++;
            giftScore[map.get(getter)]--;
            table[map.get(giver)][map.get(getter)]++;
        }

        int[] giftCount = new int[len];

        for (int i = 0; i < len; i++) {
            for (int j = 0; j < len; j++) {
                if (i == j) continue;
                if (table[i][j] > table[j][i]) giftCount[i]++;
                else if (table[i][j] == table[j][i]) {
                    if (giftScore[i] > giftScore[j]) giftCount[i]++;
                }
                
                answer = Math.max(answer, giftCount[i]);
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    lst = {}
    n = len(friends)

    for i in range(n):
        lst[friends[i]] = i #{'muzi' : 0, 'ryan' : 1, 'frodo' : 2, 'neo' : 3}

    table = [[0] * (n) for _ in range(n)] #[[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]

    #선물지수
    gift_score = [0] * n #[0,0,0,0]
    
    for gift in gifts:
        giver, getter = gift.split()
        gift_score[lst[giver]] += 1
        gift_score[lst[getter]] -= 1
        table[lst[giver]][lst[getter]] += 1

    #선물받을 개수
    gift_count = [0] * (n)
    answer = 0
    
    for i in range(n):
        for j in range(n):
            if i == j: 
                continue
            if table[i][j] > table[j][i]:
                gift_count[i] += 1
            elif table[i][j] == table[j][i]:
                if gift_score[i] > gift_score[j]:
                    gift_count[i] += 1
    
    answer = max(gift_count)
    return answer

